src/import/raft_applier.rs

```diff
@@ -0,0 +1,440 @@
+// Copyright 2023 TiKV Project Authors. Licensed under Apache-2.0.
+//! This module contains types for asynchronously applying the write batches
+//! into the storage.
+
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+
+use futures::{Stream, StreamExt};
+use kvproto::kvrpcpb::Context;
+use sst_importer::metrics::{APPLIER_ENGINE_REQUEST_DURATION, APPLIER_EVENT, IMPORTER_APPLY_BYTES};
+use tikv_kv::{with_tls_engine, Engine, WriteData, WriteEvent};
+use tikv_util::{fut, time::Instant};
+use tokio::sync::{Semaphore, SemaphorePermit};
+
+use crate::storage;
+
+async fn wait_write(mut s: impl Stream<Item = WriteEvent> + Send + Unpin) -> storage::Result<()> {
+    match s.next().await {
+        Some(WriteEvent::Finished(Ok(()))) => Ok(()),
+        Some(WriteEvent::Finished(Err(e))) => Err(e.into()),
+        Some(e) => Err(box_err!("unexpected event: {:?}", e)),
+        None => Err(box_err!("stream closed")),
+    }
+}
+
+const MAX_CONCURRENCY_PER_REGION: usize = 16;
+
+async fn acquire_semaphore(smp: &Arc<Semaphore>) -> Option<SemaphorePermit<'_>> {
+    if let Ok(pmt) = smp.try_acquire() {
+        return Some(pmt);
+    }
+    APPLIER_EVENT.with_label_values(&["raft-throttled"]).inc();
+    smp.acquire().await.ok()
+}
+
+#[derive(Clone, Default)]
+pub struct ThrottledTlsEngineWriter(Arc<Mutex<Inner>>);
+
+impl ThrottledTlsEngineWriter {
+    /// Write into the thread local storage engine.
+    ///
+    /// # Safety
+    ///
+    /// Before polling the future this returns, make sure the carrier thread's
+    /// `TLS_ENGINE_ANY` is an engine typed `E`, or at least has the same
+    /// memory layout of `E`.
+    pub unsafe fn write<E: Engine>(
```

Comment:
No need to leak unsafe, and you should make this struct pub(crate).

Reply:
I think when the future has been polled in a thread with different type of `tls_engine`, the memory safety would be broken. And in fact we cannot make a future unsafe, so I marked this function itself unsafe. I'm not sure whether it is safe to mark it safe?

Reply:
unsafe is just a marker that allows coder propagate the "unsafe" meaning outside of current module/crate (compared to a special function name like "unsafe_write", you always need to consume the "unsafe"-ness right away). For your case, as long as this function is called in a managed thread pool it will be safe. That's why I want it to be pub(crate) to avoid misusage outside of this crate.

Reply:
Make sense. However I'm sorry but I cannot fully agree with crate as the unsafe boundary. IMO though the unsafeness must be consume, why can't is be consumed at a safer place, i.e. where we are spawning the "unsafe" task to a managed thread pool? At that condition, we keep safety both in the module `raft_writer` and the crate `tikv`.(Otherwise the module `raft_writer` itself may be unsound, future developers who are adding new features into the `tikv` crate may misuse this) What do you think?

Reply:
(I haven't written many unsafe code, if my idea isn't the best practice please figure it out, so I can migrate my code style to that, thanks.

Reply:
Forgot to reply. If you think there're other uses of this other than `ImportSstService`, then it's fair to keep it unsafe.